src/goto-programs/string_refine_preprocess.cpp

```diff
@@ -0,0 +1,1343 @@
+
+\*******************************************************************/
+
+symbol_exprt string_refine_preprocesst::new_tmp_symbol(
+  const std::string &prefix, const typet &type)
+{
+  auxiliary_symbolt tmp_symbol;
+  std::ostringstream buf;
+  buf << "string_preprocess_tmp" << (next_symbol_id++) << "#" << prefix;
+  std::string name=buf.str();
+  tmp_symbol.base_name=name;
+  tmp_symbol.is_static_lifetime=false;
+  tmp_symbol.mode=ID_java;
+  tmp_symbol.name=name;
+  tmp_symbol.type=type;
+  symbol_table.add(tmp_symbol);
+  return symbol_exprt(name, type);
+}
+
+/*******************************************************************\
+
+Function: string_refine_preprocesst::new_symbol
+
+  Inputs: a name and a type
+
+ Outputs: a new symbol
+
+ Purpose: add a symbol with the given name and type to the symbol table
+          and returns it
+
+\*******************************************************************/
+
+symbol_exprt string_refine_preprocesst::new_symbol(
+  const std::string &prefix, const typet &type, bool static_lifetime)
+{
+  symbolt tmp_symbol;
+  std::ostringstream buf;
+  buf << "string_preprocess" << (next_symbol_id++) << "#" << prefix;
+  std::string name=buf.str();
+  tmp_symbol.base_name=name;
+  tmp_symbol.is_static_lifetime=static_lifetime;
+  tmp_symbol.is_state_var=static_lifetime;
+  tmp_symbol.is_type=false;
+  tmp_symbol.is_lvalue=true;
+  tmp_symbol.mode=ID_java;
+  tmp_symbol.name=name;
+  tmp_symbol.type=type;
+  symbol_table.add(tmp_symbol);
+  return symbol_exprt(name, type);
+}
+
+/*******************************************************************\
+
+Function: string_refine_preprocesst::declare_function
+
+  Inputs: a name and a type
+
+ Purpose: declare a function with the given name and type
+
+\*******************************************************************/
+
+void string_refine_preprocesst::declare_function(
+  irep_idt function_name, const typet &type)
+{
+  auxiliary_symbolt func_symbol;
+  func_symbol.base_name=function_name;
+  func_symbol.is_static_lifetime=false;
+  func_symbol.mode=ID_java;
+  func_symbol.name=function_name;
+  func_symbol.type=type;
+  symbol_table.add(func_symbol);
+  goto_functions.function_map[function_name];
+}
+
+/*******************************************************************\
+
+Function: string_refine_preprocesst::get_data_and_length_type_of_string
+
+  Inputs: an expression, a reference to a data type and a reference to a
+          length type
+
+ Purpose: assuming the expression is a java string, figure out what
+          the types for length and data are and put them into the references
+          given as argument
+
+\*******************************************************************/
+
+void string_refine_preprocesst::get_data_and_length_type_of_string(
+  const exprt &expr, typet &data_type, typet &length_type)
+{
+  assert(refined_string_typet::is_java_string_type(expr.type()) ||
+         refined_string_typet::is_java_string_builder_type(expr.type()));
```

Comment:
Hmm, I'm probably missing something: the precursor pull request appeared to carefully distinguish C and Java. Now this is Java only?